ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,265 @@
+
+The proposed name for this contract is `ValidatorSetManager`.
+
+This proposal should also help define the payload of the P-Chain notifications sent to the Subnet upon validator addition and removal.
+
+This ACP relies on concepts introduced in [ACP-73 (Warp Addressed Transactions)](https://github.com/avalanche-foundation/ACPs/pull/73) and [ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets). It depends on them to transition to an `Implementable` state.
+
+## Motivation
+
+[ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets) opens the door to managing a Subnet validator set (stored on the P-Chain) from another chain of the Avalanche network. This is achieved by sending updates to the P-Chain from a “validator manager” address on this chain. The “validator manager” address will be the only one able to issue `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` on the P-Chain to add, change the weight of, and remove validators.
+
+`Warp Addressed Transactions` (introduced in [ACP-73](https://github.com/avalanche-foundation/ACPs/pull/73)) could be used to send the validator set updates to the P-Chain.
+
+Upon addition or removal of a validator, the P-Chain is expected to send a Warp message back to the Subnet. The Subnet needs a way to interpret this message and trigger appropriate actions. These notification messages have not been specified to date.
+
+Given these assumptions and the fact that most of the active blockchains on Avalanche mainnet are EVM-based (especially the C-Chain, on which we can expect most of the Subnet “validator manager” addresses will be hosted for security reasons), we propose defining a reference implementation for a Solidity smart contract that can:
+
+1. Hold relevant information about the current Subnet validator set, as well as historical information
+2. Send validator set updates by issuing transactions defined in ACP-77 to the P-Chain
+3. Correctly update the validator set by interpreting notification messages received from the P-Chain
+4. Be easily extended with custom “security modules” to implement any security model
+
+Having an audited and open-source reference implementation freely available will contribute to lowering the cost of launching Subnets on Avalanche.
+
+Once deployed, this contract will be used as the `address` in the `SetSubnetValidatorManagerTx`.
+
+## Specification
+
+### Storage
+
+The `ValidatorSetManager` needs to store all this data to operate correctly.
+
+**Note:** This is pseudo-code that might not be 100% syntactically correct.
+
+```solidity
+// The ID of the Subnet that uses this contract to manage its validator set
+bytes32 subnetID;
+
+// The current Subnet validator set (list of NodeIDs)
+bytes32[] subnetCurrentValidatorSet;
```

Comment:
It will be 32 bytes since Subnet Validators will exclusively use Ed25519 public keys

Reply:
Yes, I took ACP-77 as reference where the `nodeID` is 32 bytes.

Reply:
Actually, I am a bit confused here. NodeIDs will always be 20 bytes no? Even for ACP-77.

Reply:
No it will be 32 bytes to house the Ed25519 public key. ACP-20 includes details on how this will be done: https://github.com/avalanche-foundation/ACPs/blob/218b62f4dd907f80f5991455dddcf72d59f33bb6/ACPs/20-ed25519-p2p/README.md

Reply:
Indeed I totally forgot about this ACP. Any timeline for implementation and rollout?